Source/Core/Core/Boot/Boot_BS2Emu.cpp

```diff
@@ -150,75 +270,128 @@ bool CBoot::RunApploader(Core::System& system, const Core::CPUThreadGuard& guard
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_CLOSE_POINTER, 4, "AppLoaderClosePointer",
+                            Common::Symbol::Type::Data);
+
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_HLE_AFTER_ENTRY, 4, "HLEAppLoaderAfterEntry");
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_HLE_AFTER_INIT, 4, "HLEAppLoaderAfterInit");
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_HLE_AFTER_MAIN, 4, "HLEAppLoaderAfterMain");
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_HLE_AFTER_CLOSE, 4, "HLEAppLoaderAfterClose");
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_HLE_REPORT, 4, "HLEAppLoaderReport");
+
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_BASE, 4, "DolphinApploaderRunner");
+  const u32 remainder_size = u32(APPLOADER_RUNNER_ASM.size() * sizeof(u32)) -
+                             (APPLOADER_RUNNER_RUN_ADDR - APPLOADER_RUNNER_BASE);
+  g_symbolDB.AddKnownSymbol(guard, APPLOADER_RUNNER_RUN_ADDR, remainder_size,
+                            "DolphinApploaderRunner");
+  g_symbolDB.Index();
+  Host_NotifyMapLoaded();
 
   auto& ppc_state = system.GetPPCState();
-  auto& mmu = system.GetMMU();
+  ppc_state.gpr[31] = APPLOADER_RUNNER_BASE;
+  ppc_state.gpr[30] = *entry;
 
-  // Call iAppLoaderEntry.
-  DEBUG_LOG_FMT(BOOT, "Call iAppLoaderEntry");
-  const u32 iAppLoaderFuncAddr = is_wii ? 0x80004000 : 0x80003100;
-  ppc_state.gpr[3] = iAppLoaderFuncAddr + 0;
-  ppc_state.gpr[4] = iAppLoaderFuncAddr + 4;
-  ppc_state.gpr[5] = iAppLoaderFuncAddr + 8;
-  RunFunction(system, *entry);
-  const u32 iAppLoaderInit = mmu.Read_U32(iAppLoaderFuncAddr + 0);
-  const u32 iAppLoaderMain = mmu.Read_U32(iAppLoaderFuncAddr + 4);
-  const u32 iAppLoaderClose = mmu.Read_U32(iAppLoaderFuncAddr + 8);
-
-  // iAppLoaderInit
-  DEBUG_LOG_FMT(BOOT, "Call iAppLoaderInit");
-  PowerPC::MMU::HostWrite_U32(guard, 0x4E800020, 0x81300000);  // Write BLR
-  HLE::Patch(system, 0x81300000, "AppLoaderReport");           // HLE OSReport for Apploader
-  ppc_state.gpr[3] = 0x81300000;
-  RunFunction(system, iAppLoaderInit);
+  // Run our apploader runner code
+  ppc_state.pc = APPLOADER_RUNNER_BASE;
 
-  // iAppLoaderMain - Here we load the apploader, the DOL (the exe) and the FST (filesystem).
-  // To give you an idea about where the stuff is located on the disc take a look at yagcd
-  // ch 13.
-  DEBUG_LOG_FMT(BOOT, "Call iAppLoaderMain");
+  return true;
+}
 
-  ppc_state.gpr[3] = 0x81300004;
-  ppc_state.gpr[4] = 0x81300008;
-  ppc_state.gpr[5] = 0x8130000c;
+void CBoot::HLEAppLoaderAfterEntry(const Core::CPUThreadGuard& guard)
+{
+  auto& system = Core::System::GetInstance();
+  auto& mmu = system.GetMMU();
 
-  RunFunction(system, iAppLoaderMain);
+  const u32 init = mmu.HostRead_U32(guard, APPLOADER_RUNNER_INIT_POINTER);
+  const u32 main = mmu.HostRead_U32(guard, APPLOADER_RUNNER_INIT_POINTER);
+  const u32 close = mmu.HostRead_U32(guard, APPLOADER_RUNNER_INIT_POINTER);
```

Comment:
Typo, should use `APPLOADER_RUNNER_MAIN_POINTER` and `APPLOADER_RUNNER_CLOSE_POINTER`.